### PR TITLE
fix(select): objects as value fix

### LINKED ISSUE #6600: Select (With Object): First open, the value is not selected
### Description

The first time that we open the Select, we cannot see the option selected.
By the way, do we really need to use a useRef in this case ?

### Reproduction

https://carbon.sage.com/?path=/docs/select--with-object-as-value

### Steps to reproduce

After loading the page:
- Open the select
- No value are loaded (Normally Green need to be selected)

### JIRA ticket numbers (Sage only)

_No response_

### Suggested solution

_No response_

### Carbon version

126.5.0

### Design tokens version

4.23.0

### Relevant browsers

Chrome

### Relevant OSs

MacOS

### Additional context

_No response_

### Confidentiality

- [X] I confirm there is no confidential or commercially sensitive information included.

### PR DESCRIPTION
**Pull request by nuria1110**
fix #6600

### Proposed behaviour

When values for options in the `Select` component are set to objects, `select-list` is not able to compare object values and does not highlight the selected option as it does not find it in the children values. Added a shallow equality check to be able to compare property values instead of object references. 

### Current behaviour

'Object as value' story for `Select` components does not highlight the preselected value. 

### Checklist

- [x] Commits follow our style guide
- [x] Related issues linked in commit messages if required
- [ ] Screenshots are included in the PR if useful
- [ ] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Playwright automation tests added or updated if required
- [x] Storybook added or updated if required
- [ ] Translations added or updated (including creating or amending translation keys table in storybook) if required
- [ ] Typescript `d.ts` file added or updated if required
- [ ] Related docs have been updated if required

#### QA

- [ ] Tested in provided StackBlitz sandbox/Storybook
- [ ] Add new Playwright test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### CODE REVIEW

File: src/components/select/select-list/select-list.component.tsx [hunk 1]
```diff
@@ -189,12 +189,33 @@ const SelectList = React.forwardRef(
...
+    ) {
+      const keysA = Object.keys(objA);
+
+      for (const key of keysA) {
+        if (objA[key] !== objB[key]) {
+          return false;
+        }
+      }
+      return true;
+    }
```
Comment (DipperTheDan) [inline src/components/select/select-list/select-list.component.tsx:204]: **suggestion(non-blocking):** I _think_ you can do this with the `.every()` JS method. I tried it locally and it seemed to still work and all the tests (unit and PW) passed. I'm happy if you want to leave it the way you've done it as it works and it's easy to read. This is what I did though:

```
    function shallowEqual(
      objA: Record<string, unknown>,
      objB: Record<string, unknown>
    ): boolean {
      const keysA = Object.keys(objA);

      return keysA.every((key) => objA[key] === objB[key]);
    }
```

File: src/components/select/select-list/select-list.spec.tsx [hunk 2]
```diff
@@ -721,7 +780,7 @@ describe("SelectList", () => {
       );
     });
 
-    it("and is in multiColum mode, then a Loader Component should be rendered", () => {
+    it("and is in multiColumn mode, then a Loader Component should be rendered", () => {
```
Comment (DipperTheDan) [inline src/components/select/select-list/select-list.spec.tsx:783]: **praise:** Well spotted.

File: src/components/select/simple-select/simple-select.pw.tsx [hunk 3]
```diff
@@ -586,7 +586,23 @@ test.describe("SimpleSelect component", () => {
...
+    mount,
+    page,
+  }) => {
+    await mount(<SimpleSelectObjectAsValueComponent />);
+
+    const inputElement = getDataElementByValue(page, "input");
+    await expect(inputElement).toHaveValue("Green");
+    await selectText(page).click();
+    const optionElement = selectOptionByText(page, "Green");
+    await expect(optionElement).toHaveCSS(
```
Comment (DipperTheDan) [inline src/components/select/simple-select/simple-select.pw.tsx:599]: **nitpick:** We've been putting a line break between the body of the test and the `expect()`.

```suggestion
    const optionElement = selectOptionByText(page, "Green");
    
    await expect(optionElement).toHaveCSS(
```

File: src/components/select/filterable-select/filterable-select.pw.tsx [hunk 4]
```diff
@@ -720,7 +720,23 @@ test.describe("FilterableSelect component", () => {
...
+    mount,
+    page,
+  }) => {
+    await mount(<FilterableSelectObjectAsValueComponent />);
+
+    const inputElement = getDataElementByValue(page, "input");
+    await expect(inputElement).toHaveValue("Green");
+    await dropdownButton(page).click();
+    const optionElement = selectOptionByText(page, "Green");
+    await expect(optionElement).toHaveCSS(
```
Comment (DipperTheDan) [inline src/components/select/filterable-select/filterable-select.pw.tsx:733]: **nitpick:** We've been putting a line break between the body of the test and the `expect()`.

```suggestion
    const optionElement = selectOptionByText(page, "Green");
    
    await expect(optionElement).toHaveCSS(
```

File: src/components/select/select-list/select-list.component.tsx (hunk 1 above)
Comment (edleeks87) [reply src/components/select/select-list/select-list.component.tsx]: you can short circuit it faster if you invert the logic and use `some`:

```
 function shallowEqual(
      objA: Record<string, unknown>,
      objB: Record<string, unknown>
    ): boolean {
      const keysA = Object.keys(objA);

      return !keysA.some((key) => objA[key] !== objB[key]);
    }

**Review by edleeks87 — DISMISSED** [review]

**Review by DipperTheDan — APPROVED** [review]

**Review by ZhuoyuJin — DISMISSED** [review]

**Review by edleeks87 — APPROVED** [review]

**Review by DipperTheDan — APPROVED** [review]

**Review by ZhuoyuJin — APPROVED** [review]

**Pull request merged**